### PR TITLE
The Catalina Merge

### PR DESCRIPTION
**Pull request by millenomi**
This patch brings swift-corelibs-foundation closer to behavior parity with released versions of macOS Catalina and aligned Apple releases. Most of these changes are under-the-hood bug fixes or performance improvements, but a few may affect the behavior of your applications. In particular:

 - The deprecation of NSURL.parameterString is now also effective in Swift for Linux, with similar effects. This brings URL parsing in line with Catalina. Note that Swift for Linux does not support link checks, since there is no binary compatibility guarantee on Linux; this change will be effective to all application code running on Swift for Linux 5.3 or later.

 - Certain CFXML… types that have been deprecated on macOS are now removed, along with their associated implementation: CFXMLInputStream, CFXMLNode, CFXMLParser. Modern code should be using the FoundationXML module (or, on Darwin, the XMLDocument and related classes in Foundation) to parse documents.

### CODE REVIEW

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
https://dev.azure.com/compnerd/swift-build/_build/results?buildId=31292&view=results for a run on Windows

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

File: CoreFoundation/Base.subproj/CFPlatform.c [hunk 1]
```diff
@@ -1420,36 +1476,18 @@ _CFThreadSpecificKey _CFThreadSpecificKeyCreate() {
 
 CFTypeRef _Nullable _CFThreadSpecificGet(_CFThreadSpecificKey key) {
 #if TARGET_OS_WIN32
-    _CFThreadSpecificData *data = (_CFThreadSpecificData *)FlsGetValue(key);
-    if (data == NULL) {
-        return NULL;
-    }
-    return data->value;
+  return (CFTypeRef)FlsGetValue(key);
```
Comment (xwu) [inline CoreFoundation/Base.subproj/CFPlatform.c:1479]:
```suggestion
    return (CFTypeRef)FlsGetValue(key);
```

Comment (millenomi) [reply CoreFoundation/Base.subproj/CFPlatform.c:1479]: I've seen a few like this. Especially wrt the implementation of TSDs on Windows, @compnerd, let me know if it looks bad?

Comment (millenomi) [reply CoreFoundation/Base.subproj/CFPlatform.c:1479]: My mental map of The Whole Of Foundation gets a little murky around the Windows bits.

File: CoreFoundation/Base.subproj/ForFoundationOnly.h [hunk 2]
```diff
@@ -121,44 +142,53 @@ typedef struct {
...
 #define NSSTRING_POSSIBLE_RANGEERROR(range, len)     \
-    if (__CFStringNoteErrors()) {       \
+    do {       \
         static bool warnonce = false;   \
         if (!warnonce) {                \
             warnonce = true;            \
             CFLog(kCFLogLevelWarning, CFSTR("*** %@: Range {%lu, %lu} out of bounds; string length %lu. This will become an exception for apps linked after 10.10 and iOS 8. Warning shown once per app execution."), __CFExceptionProem((id)self, _cmd), (unsigned long)range.location, (unsigned long)range.length, (unsigned long)len);        \
     }   \
-}
+} while(0)
```
Comment (xwu) [inline CoreFoundation/Base.subproj/ForFoundationOnly.h:175]: Nit: the indentation is off here.

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

File: CoreFoundation/Parsing.subproj/CFBinaryPList.c [hunk 3]
```diff
@@ -1221,19 +1216,27 @@ CF_PRIVATE bool __CFBinaryPlistCreateObjectFiltered(const uint8_t *databytes, ui
...
-	    *plist = CFNumberCreate(allocator, kCFNumberSInt128Type, &val);
-	} else {
-	    *plist = CFNumberCreate(allocator, kCFNumberSInt64Type, &bigint);
-	}
-	// these are always immutable
-	if (objects && *plist) {
-	    CFDictionarySetValue(objects, (const void *)(uintptr_t)startOffset, *plist);
-	}
-	return (*plist) ? true : false;
+        if (outPlistTypeID) *outPlistTypeID = _kCFRuntimeIDCFNumber;
```
Comment (xwu) [inline CoreFoundation/Parsing.subproj/CFBinaryPList.c:1219]: Nit: indentation appears off here and below.

File: CoreFoundation/Parsing.subproj/CFBinaryPList.c [hunk 4]
```diff
@@ -1248,12 +1251,19 @@ CF_PRIVATE bool __CFBinaryPlistCreateObjectFiltered(const uint8_t *databytes, ui
 	    CFSwappedFloat32 swapped32;
 	    memmove(&swapped32, ptr, 4);
 	    float f = CFConvertFloat32SwappedToHost(swapped32);
-	    *plist = CFNumberCreate(allocator, kCFNumberFloat32Type, &f);
-	    // these are always immutable
-	    if (objects && *plist) {
-		CFDictionarySetValue(objects, (const void *)(uintptr_t)startOffset, *plist);
-	    }
-	    return (*plist) ? true : false;
+            if (outPlistTypeID) *outPlistTypeID = _kCFRuntimeIDCFNumber;
```
Comment (xwu) [inline CoreFoundation/Parsing.subproj/CFBinaryPList.c:1254]: Nit: likewise indentation appears off here and below, and it appears pervasively throughout the rest of the file.

File: CoreFoundation/Base.subproj/ForFoundationOnly.h (hunk 2 above)
Comment (millenomi) [reply CoreFoundation/Base.subproj/ForFoundationOnly.h:175]: I'm gonna note these down and work on them later, for internal back-and-forth merging reasons.

**Comment by millenomi** [comment]
@swift-ci please test

File: CoreFoundation/RunLoop.subproj/CFMessagePort.c [hunk 5]
```diff
@@ -837,23 +822,24 @@ static void __CFMessagePortReplyCallBack(CFMachPortRef port, void *msg, CFIndex
...
 	    uintptr_t data_extent = (uintptr_t)((uint8_t *)&(MSGP_INFO(replymsg, bytes)) + byteslen);
             if (byteslen < 0) byteslen = 0; // from here on, treat negative same as zero -- this is historical behavior: a NULL return from the callback on the other side results in empty data to the original requestor
 	    if (0 <= byteslen && data_extent <= msgp_extent) {
 		reply = CFDataCreate(kCFAllocatorSystemDefault, MSGP_INFO(replymsg, bytes), byteslen);
-	    } else {
-		reply = (void *)~0;	// means NULL data
-	    }
-	} else {
-//#warning CF: should create a no-copy data here that has a custom VM-freeing allocator, and not vm_dealloc here
+            } else {
```
Comment (xwu) [inline CoreFoundation/RunLoop.subproj/CFMessagePort.c:833]: Nit: indentation consistency problem here also, it appears.

File: CoreFoundation/RunLoop.subproj/CFRunLoop.c [hunk 6]
```diff
@@ -2419,15 +2326,19 @@ static Boolean __CFRunLoopDoTimers(CFRunLoopRef rl, CFRunLoopModeRef rlm, uint64
...
     
     for (CFIndex idx = 0, cnt = timers ? CFArrayGetCount(timers) : 0; idx < cnt; idx++) {
         CFRunLoopTimerRef rlt = (CFRunLoopTimerRef)CFArrayGetValueAtIndex(timers, idx);
         Boolean did = __CFRunLoopDoTimer(rl, rlm, rlt);
         timerHandled = timerHandled || did;
     }
-    
-    if (timers) CFRelease(timers);
-    
+        if (timers) CFRelease(timers);
```
Comment (xwu) [inline CoreFoundation/RunLoop.subproj/CFRunLoop.c:2337]:
```suggestion
    if (timers) CFRelease(timers);
```

File: CoreFoundation/String.subproj/CFString.c [hunk 7]
```diff
@@ -6420,7 +6495,7 @@ reswtch:switch (ch) {
 	    if (spec->size != CFFormatSize16) spec->size = CFFormatSize8;
 	    return true;
 	case 'n':		/* %n is not handled correctly; for Leopard or newer apps, we disable it further */
-	    spec->type = 1 ? CFFormatDummyPointerType : CFFormatPointerType;
+            spec->type = CFFormatDummyPointerType;
```
Comment (xwu) [inline CoreFoundation/String.subproj/CFString.c:6498]:
```suggestion
          spec->type = CFFormatDummyPointerType;
```

File: CoreFoundation/String.subproj/CFString.c [hunk 8]
```diff
@@ -3463,359 +3507,517 @@ enum {
...
+// Given an index, attempts to return the range of the containing element of Grapheme Cluster Boundary Rule GB11:
+//
+//   \p{Extended_Pictographic} (Extend* ZWJ \p{Extended_Pictographic})*
+//
+// Specifically, this will attempt to match either the lone starting \p{Extended_Pictographic} if the index corresponds to it, or a singular instance of (Extend* ZWJ \p{Extended_Pictographic}) which contains the `index`.
+// For instance, the string @"stuff...👩‍❤️‍💋‍👨stuff..." is segmented this way as:
+//
+//                   👩             ❤    Var_Sel           💋             👨
+//   ┌──────────┐│┌───────┐┌─────┐┌──────┐┌──────┐┌─────┐┌───────┐┌─────┐┌───────┐│┌──────────┐
+//   │ stuff... │ │ 1F469 ││ ZWJ ││ 2764 ││ FE0F ││ ZWJ ││ 1F48B ││ ZWJ ││ 1F468 │ │ stuff... │
```
Comment (xwu) [inline CoreFoundation/String.subproj/CFString.c:3597]: These diagrams are fantastic!

File: CoreFoundation/StringEncodings.subproj/CFStringEncodingConverter.c [hunk 9]
```diff
@@ -607,6 +607,7 @@ static const _CFEncodingConverter *__CFGetConverter(uint32_t encoding) {
 #warning This case must match __defaultEncoding value defined in CFString.c
 	case kCFStringEncodingISOLatin1: commonConverterSlot = (const _CFEncodingConverter **)(&(commonConverters[1])); break;
 #endif
+        case kCFStringEncodingInvalidId: return NULL;
```
Comment (xwu) [inline CoreFoundation/StringEncodings.subproj/CFStringEncodingConverter.c:610]: Nit: space-tab mismatch here.

File: CoreFoundation/URL.subproj/CFURL.c [hunk 10]
```diff
@@ -3756,7 +3715,7 @@ static Boolean decomposeToRFC1808(CFURLRef url, CFURLComponentsRFC1808 *componen
     } else {
         components->port = kCFNotFound;
     }
-    components->parameterString = _retainedComponentString(url, HAS_PARAMETERS, false, false);
+        components->parameterString = NULL;
```
Comment (xwu) [inline CoreFoundation/URL.subproj/CFURL.c:3718]:
```suggestion
    components->parameterString = NULL;
```

File: CoreFoundation/URL.subproj/CFURLAccess.c [hunk 11]
```diff
@@ -629,7 +629,9 @@ static Boolean _CFDataURLCreateDataAndPropertiesFromResource(CFAllocatorRef allo
 	if (! parseDataRequestURL(url, &data, &mimeType, &textEncodingName)) {
 		if (errorCode)
 			*errorCode = kCFURLUnknownError;
-		*fetchedData = NULL;
+                if (fetchedData) {
```
Comment (xwu) [inline CoreFoundation/URL.subproj/CFURLAccess.c:632]: Nit: space-tab mismatch here.

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
The patch passes on Linux and Mac but breaks the Windows build; there will be some iteration.

**Comment by lxbndr** [comment]
This reverts #2764. I'm afraid NotificationQueue could be affected, as well as Operation/OperationQueue.

**Comment by millenomi** [comment]
Thank you for catching that. I’ll un-revert.

**Comment by xwu** [comment]
For ease of reviewing, would it be possible to commit the ICU upgrade separately? It would at least make the GitHub interface slightly more usable.

**Comment by millenomi** [comment]
I haven't tested the ICU upgrade with the previous version of the code, and I don't know that I want to put in the time to do that.

I understand this is a large patch, but extricating it would add more to what has been so far already ~1 work-hours-month-person of work that I need to coordinate with some expedition, and I'd rather not unless the need is _really strong_.

**Comment by millenomi** [comment]
(The ICU parts are showing collapsed for me? I'm not sure how their presence impacts usability. While you can review them, I'm not going to modify them: this is just a straight import of the headers generated by compiling Apple ICU from https://opensource.apple.com/source/ICU/ICU-64252.0.1/.)

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by xwu** [comment]
@millenomi Yeah, I mention this thought precisely because there is no point in reviewing the ICU parts. Having not interfaced with the library extensively myself, I don't have a sense of its forward compatibility guarantees. If it's essentially a drop-in replacement, then perhaps there's some net benefit to doing that separately. If it's much more involved than that, then there's certainly no point in extricating it.

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
I'm going to avoid starting tests when I'm doing Windows runs.

**Comment by millenomi** [comment]
https://dev.azure.com/compnerd/swift-build/_build/results?buildId=31349&view=results

**Comment by millenomi** [comment]
This one needs a test.

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
https://dev.azure.com/compnerd/swift-build/_build/results?buildId=31352&view=logs&s=2bc04634-cf71-5486-fada-47995d15befc

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
https://dev.azure.com/compnerd/swift-build/_build/results?buildId=31353&view=results

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
https://dev.azure.com/compnerd/swift-build/_build/results?buildId=31368&view=logs&s=2bc04634-cf71-5486-fada-47995d15befc

**Comment by millenomi** [comment]
This is going to be merged today barring last-minute objections.

**Comment by xwu** [comment]
Last minute objection: Unless I misread, I think your fix is still not quite right after the amendment. The wrapper function inverts the logic of `os_mul_overflow`, but on non-Win32 platforms is an alias of `os_mul_overflow`...

**Comment by millenomi** [comment]
Thanks for checking!

**Comment by millenomi** [comment]
Merging this.

**Comment by buttaface** [comment]
Looks like [this broke](https://github.com/apple/swift-package-manager/pull/2740#issuecomment-626922147) the [linux CI](https://ci.swift.org/job/swift-syntax-PR-Linux/23/console), because of [this line](https://github.com/apple/swift-corelibs-foundation/pull/2782/files#diff-77656c27b6ca4fa4e9211ac6241a539fR3552):
```
swift-corelibs-foundation/CoreFoundation/String.subproj/CFString.c:3552:100: error: use of undeclared identifier 'UCHAR_EMOJI_MODIFIER'
```

**Pull request merged**